cardano-api/src/Cardano/Api/Fees.hs

```diff
@@ -837,24 +874,49 @@ makeTransactionBodyAutoBalance eraInMode systemstart history pparams
    era' :: CardanoEra era
    era' = cardanoEra
 
-   balanceCheck :: Lovelace -> Lovelace -> Either TxBodyErrorAutoBalance ()
-   balanceCheck minUTxOValue balance
-    | balance < 0            = Left (TxBodyErrorAdaBalanceNegative balance)
-      -- check the change is over the min utxo threshold
-    | balance < minUTxOValue = Left (TxBodyErrorAdaBalanceTooSmall balance)
-    | otherwise              = return ()
-
-   getMinUTxOValue :: ProtocolParameters
-                   -> Either TxBodyErrorAutoBalance Lovelace
-   getMinUTxOValue pparams' =
+   balanceCheck :: TxOutValue era -> Either TxBodyErrorAutoBalance ()
+   balanceCheck balance
+    | txOutValueToLovelace balance < 0 =
+        Left . TxBodyErrorAdaBalanceNegative $ txOutValueToLovelace balance
+    | otherwise =
+        checkMinUTxOValue (TxOut changeaddr balance TxOutDatumHashNone) pparams
```

Comment:
It's not a good idea to merge the `balance < minUTxOValue` check here into the general `checkMinUTxOValue`.

Note how the code now never uses the `TxBodyErrorAdaBalanceTooSmall` error constructor and so will report the wrong error message.

I actually don't see why we need to change the `balanceCheck` code at all. It does not need to be given the full `TxOutValue`. As far as I can see the `Lovelace` only was fine.

Reply:
The minUTxO value is calculated differently depending on the era, which is why I put `checkMinUTxOValue` in `balanceCheck`.